### PR TITLE
Creating a new ToggleButton GUI control

### PR DESCRIPTION
**Pull request by kintz09**
**Summary:**

[Original Pull Request](https://github.com/BabylonJS/Babylon.js/pull/9425)

This pull request creates a ToggleButton class, extended from the Rectangle class. It functions as an individual Toggle or as a toggle group. The design of the class followed the Button class, as this is just another button type with some unique properties to support the use as a toggle button.

**Features**
`isActive` property with getter/setter to define if the toggle is active/on or inactive/off.

`toActiveAnimation` and `toInactiveAnimation` functions allow the developer to easily define the toggle's properties when in an active and inactive state. These functions fire when the `isActive` property changes.

`pointerEnterActiveAnimation`, `pointerOutActiveAnimation`, `pointerDownActiveAnimation`, `pointerUpActiveAnimation`, `pointerEnterInactiveAnimation`, `pointerOutInactiveAnimation`, `pointerDownInactiveAnimation`, `pointerUpInactiveAnimation` functions allow the developer to easily define separate pointer animations for the Active and Inactive states. These were implemented following Button.ts as a guide.

`group` property allows the developer to define the group the toggle belongs to. When a Toggle in a group is toggled on, the other toggles in the group will be toggled off. This was implemented following radioButton.ts as a guide.

`onIsActiveChangedObservable` an Observable raised when the `isActive` property value changes. This was implemented with radioButton.ts as a guide.

I attached the typescript playground code I used to test the class which can also be used as an example.
[toggleButton-playground-code.txt](https://github.com/BabylonJS/Babylon.js/files/5586074/toggleButton-playground-code.txt)

[Original Discussion](https://forum.babylonjs.com/t/best-way-to-create-and-animate-custom-button-styles-like-for-a-disabled-state/15583)

### CODE REVIEW

**Comment by sebavan** [comment]
Looks pretty great, @Popov72  will make a deeper review of it, and we might wait on @deltakosh next week for the final review if necessary.

File: gui/src/2D/controls/toggleButton.ts
Comment (Popov72) [inline gui/src/2D/controls/toggleButton.ts:79]: Wrong doc (copy/paste of the previous one)

File: gui/src/2D/controls/toggleButton.ts
Comment (Popov72) [inline gui/src/2D/controls/toggleButton.ts:128]: Maybe you can get use `_getTypeName() === "ToggleButton"` instead of testing for the `group` property (in case one day other control(s) define a `group` property)?

File: gui/src/2D/controls/toggleButton.ts
Comment (Popov72) [inline gui/src/2D/controls/toggleButton.ts:115]: You test event handlers for not being undefined further down, so I guess you should also do it here.

**Review by Popov72 — CHANGES_REQUESTED** [review]
LGTM, only some small comments.

File: gui/src/2D/controls/toggleButton.ts
Comment (kintz09) [reply gui/src/2D/controls/toggleButton.ts:115]: Would an optional chaining operator be acceptable in the BabylonJS project? 

_The traditional way:_
`if (this._isActive) {
    if (this.toActiveAnimation) {
        this.toActiveAnimation();
    }
} else {
    if (this.toInactiveAnimation) {
        this.toInactiveAnimation()
    }
}`

_Using Optional Chaining:_
`if (this._isActive) {
    this.toActiveAnimation.?();
} else {
    this.toInactiveAnimation.?();
}`

From the [Mozilla's documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining), it looks safe to use, but wasn't sure if there were compatibility concerns.

Comment (sebavan) [reply gui/src/2D/controls/toggleButton.ts:115]: should be good as transpiled by ts anyway

File: gui/src/2D/controls/toggleButton.ts
Comment (kintz09) [reply gui/src/2D/controls/toggleButton.ts]: I did a rewrite of this function utilizing `_getTypeName()` and tried to simplify the logic. Hopefully the comments make sense, but if not let me know and I can give you pseudocode to review.

File: gui/src/2D/controls/toggleButton.ts
Comment (kintz09) [reply gui/src/2D/controls/toggleButton.ts:79]: This should be resolved in the latest update to this pull request.

File: gui/src/2D/controls/toggleButton.ts
Comment (kintz09) [reply gui/src/2D/controls/toggleButton.ts:115]: This should be resolved in the latest update to this pull request.

File: gui/src/2D/controls/toggleButton.ts
Comment (Popov72) [inline gui/src/2D/controls/toggleButton.ts:126]: Something like this should work:
```javacsript
if (control.typeName === "ToggleButton") {
```
as `typeName` is a property of the base class `Control`.

Comment (kintz09) [reply gui/src/2D/controls/toggleButton.ts]: Ah, that feels much cleaner. This pull request has been updated with this change.

**Comment by sebavan** [comment]
Looks overall good to me but as we are trying to merge a huge change in today, is that a problem if we wait for @deltakosh on Monday to validate the PR ???

**Comment by kintz09** [comment]
Not sure if you're asking me, but if so, there's no urgency to get this merged.

**Comment by sebavan** [comment]
Yup that was exactly my question, thanks a lot.

**Review by deltakosh — APPROVED** [review]

**Pull request merged**
